src/utils/from-ipld-node.js

```diff
@@ -0,0 +1,28 @@
+const cidToMultihash = (cid) => {
+  if (!cid) {
+    return cid
+  }
+
+  if (Array.isArray(cid)) {
+    return cid.map(cidToMultihash)
+  }
+
+  // Ensure backwards compatibility by checking if this is a multihash already
+  if (typeof cid === 'string') {
+    return cid
+  }
+
+  return cid.toBaseEncodedString()
+}
+
+const fromIpldNode = (dagNode, props) => {
+  const obj = Object.assign({}, dagNode)
```

Comment:
@satazor version 0 of ipfs-log appears to create 'dag-pb' objects so this still fails backwards-compatibility when I try to open an old orbit-db address. I'm not sure its the best way but it works if we change this to sth like the following: `const obj = dagNode.data ? Object.assign({}, JSON.parse(dagNode.data)) : Object.assign({}, dagNode)`. What do you think?

That along with the version increment @aphelionz mentioned should do the trick

Reply:
You are right absolutely right. I will take care of that and add tests as well.

Reply:
One question: If I call `fromMultihash` with a hash of a `dag-pb` and then call `toMultihash`, it should give the same hash right? If so, I think it's worth doing "proper" migration code that handles reading/writing nodes with version support.

Reply:
How would that work in a "mixed" log - i.e. one with some `v: 0` entries and some `v: 1` entries? Another way of asking is - how would the log know which multihash type it was created from?

Reply:
@aphelionz The way I see it:

- Have a `read` step given a multihash. The read function would essentially return a JSON object directly from `dag.get`. This is where we can test if a hash is a `dag-pb` and use JSON.parse or not. We may use [cid](https://github.com/multiformats/js-cid) for this or just a `typeof` check.
- Have a migration pipeline. We setup a set of functions to migrate from previous versions, e.g.: we would have functions to migrate from v0 to v1, v1 to v2 and so on. I suspect we may find a few libraries on npm that may help us with this. This is based on the `v` field that is tagged on the JSON object.
- Finally, the `fromIpldNode` and `toIpldNode` would still be called by `entry-io.js` and `log.js` but is out of scope of the "migration" pipeline.

Reply:
Hmm good point, I agree fromMultihash and toMultihash should return the same. Perhaps we should change `Entry.toMutihash` and use `DAGNode.create` and return the cid if `entry.v === 0`. What do you think?

Reply:
@shamb0t Yes I can do that but it feels ad-hoc. I would suggestion embrancing/crafting a migration pipeline that would take care of that and future changes on both reads and writes. What do you think?

Reply:
Woops missed your last comment. Migration pipeline would be great but is it not an issue that each entry needs to be signed? That means logs with many users wouldnt be able to be migrated bc incrementing the version of an entry would require new signature from each user. So I think backwards compatibility will just be for reading old entries but not updating them. Does this make sense? I may be missing something

Reply:
@satazor yea it is quite hacky but not sure we can update existing entries as they are signed

Reply:
@shamb0t Correct, new entries would be always written using the latest version they are running. Note that I may still a node running an old version and I will be writing those nodes in that version.

Nevertheless, calling `Entry.fromMultihash` followed by `toMultihash` would respect the `version` the entry is written, thus giving the same hash. But calling, `Entry.create` would be creating an entry with the most recent version always.

Reply:
> @satazor yea it is quite hacky but not sure we can update existing entries as they are signed

Yep, we shouldn't update the existing entries. The migration I'm describing is on reads, thus is like a `runtime` migration without affecting the storage. The migration should also read writes only with the goal of giving the same multihash when calling `Entry.fromMultihash` followed by toMultihash`

Reply:
(updated my last comment)

Reply:
Ah maybe I am misunderstanding what you mean by migrating here? Not updating existing entries?

Reply:
One important thing: if a log instance is trying to read/write a node in which the version is not supported, it should fail with an error. This is the case where an outdated instance is being synched with a newer one.

Reply:
Perhaps "migration" is a bad name because it somewhat implies changing the storage. Do you have a better name?

Reply:
> Perhaps "migration" is a bad name because it somewhat implies changing the storage. Do you have a better name?

I guess not :p I think we're in agreement about desired behaviour but not sure Im grasping where this step should be and how its different to fromMultihash. What do you mean by migrate from v0 to v1 here?

> * Have a migration pipeline where we setup a set of functions to migrate from previous versions, e.g.: we would have functions to migrate from v0 to v1, v1 to v2 and so on.

Reply:
Do you mean fromMultihash should be called within this pipeline and not do the version check itself?

Reply:
> I guess not :p I think we're in agreement about desired behaviour but not sure Im grasping where this step should be and how its different to fromMultihash. What do you mean by migrate from v0 to v1 here?

I went to the whiteboard and wrote something to help you (and me):

![image](https://user-images.githubusercontent.com/1017236/50649185-1df9a100-0f75-11e9-8815-6e3ee6f62f81.png)

The "writing" would be similar but backwards.

This procedure will allows us to change fields or the data-structure of the entries easily. It would also allow us to change the codec in the future, in case we are not happy with `dag-cbbor`. While this is not complex to code, perhaps I suggest to keep this PR simple and and address this on #211.

I've changed the code to handle `dag-pb` nodes, will add a few tests.

Reply:
Expanding on this, the `migration` pipeline would be composed by something like this:

```js
{
    '2': {
        up(data) { return transformedData },
        down(data) { return transformeData },
        normalizeIpldLinks(data) { return dataWithLinksNormalizedToMultihashes },
     },
     '3': ...
}
```

Where:

- `up()` would transform the JSON `data` to their next version
- `down` would transform the JSON `data` to their previous version
- `normalizeIpldLinks` would be called by the step `1` (just once) so that IPLD links are transformed to regular multihashes (strings)


}

Reply:
> While this is not complex to code, perhaps I suggest to keep this PR simple and and address this on #211.

While I said this, the `fromMultihash` and `toMultihash` is still a problem in this PR. The migration scheme I detailed above would solve this or we may do a check like @shamb0t suggested in the `toMultihash` functions for now.

Reply:
We are very close, @satazor! I would like @hardcore to weigh in but I would personally implore you to finish that last piece and achieve full backwards compatability.

Alternatively, it might be ok if fromMultihash _accepts_ either type but toMultihash only returns cids.

Reply:
@aphelionz fromMultihash isn't the problem, toMultihash is. We could add a optional `version` argument to it, but it would feel strange, specially since the entries do not need it. I will look into what needs to be changed to add a `v` field to the `log.js` and `log-io.js` but I'm afraid it would require a few changes, that's why I suggested postponing adding support for IPLD in them later on.

Lets see what @haadcode has to say.

Reply:
Thanks for the ping 👍 Reading through the conversation, couple of thoughts come to mind:
- We should try to keep the solution as simple (=lightweight) as possible for now
- I like the idea of a "migration pipeline" as this is prolly not the last time we'll need to provide migrations
- That ^ said, I think it would be helpful to spend more time to figure out the full solution to make sure it's the right solution. it sounds to me this can be done at a later point and doesn't have to be part of this PR. It also reminds me that we'll probably need something similar, sooner or later, for the stores in OrbitDB and ideally a solution would facilitate both use cases (log and stores).
- We should (try to) keep all "log information" in the entries: this is something we looked into very early on and decided to not add "meta info" for the log itself and rather consider the structure of the log to be defined by the entries + public API of the log (ie. the functions you can call). This keeps the abstractions and data structures simple imo.
- We should, if possible, make it so that a log can contain mixed (Entry) versions. Ie. read a v0 log, but when appending to it, it'll be in v1 format, eg. `(e1, v0)->(e2, v0)->(e3, v1)`.
- Re. the `to/FromMultihash`, agreed that the semantics should be consistent and it is indeed a bit tricky: on one hand it would make sense that `read` (=fromMultihash) is backwards compatible in that it can take either a multihash or CID, and that `write` (=toMultihash) would return the same hash. On the other hand, it would make sense that `write` returns the _same kind of hash for every log_ and given we're going from multihashes to CIDs in the entries, it would make sense to me that the `write` returns a CID.

It sounds to me that the big question is whether, when persisting a log to IPFS, should return a multihash or a CID. The options are to 1) write a full-fledged migration pipeline or 2) figure out the right semantics for `to/fromMultihash` between the different versions. I think there's a third option, which would be to add `toCID` and `fromCID` functions that semantically do the same as `to/fromMultihash` but work on CIDs. Further, we could add a `console.warn("This function will be deprecated in the next release")` to `to/fromMultihash` to encourage users to switch. This would probably still require us to provide backwards compat in the `fromCID` part (maybe?), but from semantics perspective it would be consistent. 

What do you think, would that be a good approach and would it cover all the cases discussed above?

Reply:
> We should try to keep the solution as simple (=lightweight) as possible for now

Agree.

> I like the idea of a "migration pipeline" as this is prolly not the last time we'll need to provide migrations
That ^ said, I think it would be helpful to spend more time to figure out the full solution to make sure it's the right solution. it sounds to me this can be done at a later point and doesn't have to be part of this PR. It also reminds me that we'll probably need something similar, sooner or later, for the stores in OrbitDB and ideally a solution would facilitate both use cases (log and stores).

Agree, I can later on formalize my proposal in #211 

> We should (try to) keep all "log information" in the entries: this is something we looked into very early on and decided to not add "meta info" for the log itself and rather consider the structure of the log to be defined by the entries + public API of the log (ie. the functions you can call). This keeps the abstractions and data structures simple imo.

> Re. the to/FromMultihash, agreed that the semantics should be consistent and it is indeed a bit tricky: on one hand it would make sense that read (=fromMultihash) is backwards compatible in that it can take either a multihash or CID, and that write (=toMultihash) would return the same hash. On the other hand, it would make sense that write returns the same kind of hash for every log and given we're going from multihashes to CIDs in the entries, it would make sense to me that the write returns a CID.

Making `toMultihash` always return a CID would indeed simplify. Is there any reason, at this point, to have them consistent? Will it somehow make the log behave badly? Looking at the code, it seems it doesn't. As I said. making `fromMultihash` consistent for entries is easy, but not for logs because we would have to tag them with the version as well. We got 3 options:

1. Drop the consistency of `fromMultihash` and `toMultihash`; the `toMultihash` would always use the latest codec (dag-cbor) and return a cid (is still a string but more complete)
2. Add a `version` argument to the Log constructor that defaults to the latest version. The `fromMultihash` function would create a Log instance, passing `version` = `data.version || 0`. Then, the `toMultihash` would have the version to decide to write a `dag-pb` or a `dag-cbor`.
3. When calling `toMultihash`, analyse the head entries in order to calculate the higher one. This would be used to decide which codec to use. But I think this would require further IO?

What do you suggest going for?

>  I think there's a third option, which would be to add toCID and fromCID functions that semantically do the same as to/fromMultihash but work on CIDs. Further, we could add a console.warn("This function will be deprecated in the next release") to to/fromMultihash to encourage users to switch.

I like this suggestion, perhaps we could do that in another PR after we figure this out.

Reply:
Thanks for the whiteboarding/explanation @satazor, very helpful and I see this is more future-proof. With that, I agree we should try to keep this PR as simple as possible. I've successfully opened an old orbit-db address using ipfs-log from this PR :tada: if we agree to address the migration pipeline separately, is this ready for final review/merge?

@haadcode re: "mixed-logs" it makes sense that appending would use the latest version, but we should discuss how a peer knows it needs to update....it will only know once it receives an entry for a version it doesnt understand, and as @satazor said it should error out in such case, before receiving such an entry it would still be able to append with a previous version

Reply:
@shamb0t 

> I've successfully opened an old orbit-db address using ipfs-log from this PR 🎉 if we agree to address the migration pipeline separately, is this ready for final review/merge?

Not yet, we are still deciding if calling `fromMultihash` followed by a `toMultihash` should be consistent or not. Based on the decision, I might need to do some changes and add a few tests.

Reply:
@shamb0t I forgot to mention that solution as well, I updated my comment above and changed option 3 to be to/fromCID.

Reply:
I like introducing to/fromCID. It's not _so_ much work to change or remove it in the future, and then we can keep to/fromMultihash around. All four functions will do what they say - unless there's a more specific term we can name to/fromMultihash to that I'm not thinking of right now.

Reply:
If we go for the `to/fromCID` route, should we change every occurrence of hash, multihash to cid in all the source code and tests? If so, we can't change the `hash` property to `cid` in entries, otherwise the signatures would be invalid. There's also `Log.findTailHashes` but I think it's not part of the public API, so we can rename it.

Reply:
@haadcode could you clarify if `log.fromEntry` and `findTailHashes` are part of the public API?

Reply:
I believe `fromEntry` is but `findTailHashes` is not.

Reply:
Ooops, I mean `log.fromEntryHash`.

Reply:
`log.fromEntryHash` I don't think is part of the public API

Reply:
@satazor I feel changing all occurrences is unnecessary unless we do so in the whole codebase, entry included or it would be confusing. This would be quite a lot of work imo. Do you think its confusing to return a multihash from the `hash` property? The multihash from the cid is still what we're storing right? Or do I misunderstand?

Reply:
Yes the `hash` will have the cid from now on while old entries still have the multihash. Both are strings and reading will be backwards compatible.

Reply:
@shamb0t I agree with you. We can change `hash` to `cid` and maintain compatibility, including signatures verification. Moreover, we can keep the `hash` property around, but hidden, by making it non-enumerable.

Reply:
I'm give `toCID` and `fromCID` a shot, see my comment below.